atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -493,7 +493,7 @@ protected void checkGetPreconditions(TableReference tableRef) {
                 if (raw.isEmpty()) {
                     return endOfData();
                 }
-                Map<Cell, byte[]> post = new LinkedHashMap<>();
+                SortedMap<Cell, byte[]> post = new TreeMap<>();
                 getWithPostFiltering(tableRef, raw, post, Value.GET_VALUE);
```

Comment:
I've come across cases in hot code where building a TreeMap from almost-sorted data can cause significant perf issues (since the red-black tree is constantly rebalancing). It _may_ be worth collecting these in a list and then sorting at the end - java default to TimSort which is particularly good on almost-sorted data. Of course whether this is worthwhile depends on the scale of the data, how hot the code is, and how out-of-order the normal case is.

Reply:
How large would the result set need to be for these perf issues to become apparent? We have a pretty decent suite of performance tests, and could add a test case to expose the large-but-almost-sorted case.

Reply:
Quick and dirty benchmark:
```
    public static void main(String[] args) {
        for (int i = 10; i < 1000000000; i *= 10) {
            runTest(i);
            System.out.println();
        }
    }

    static class CountingComparator implements Comparator<String> {
        int comparisons = 0;

        @Override
        public int compare(String o1, String o2) {
            comparisons++;
            return o1.compareTo(o2);
        }
    }

    private static void runTest(int howMany) {
        List<String> list = new ArrayList<>(howMany);
        for (int i = 0; i < howMany; i++) {
            list.add(RandomStringUtils.random(20));
        }

        list.sort(null);
        // now make 'almost sorted'
        int numSwaps = (int) Math.ceil(Math.log(howMany));
        for (int i = 0; i < numSwaps; i++) {
            swapRandom(list);
        }
        System.out.println("Testing with " + howMany + "; num swaps: " + numSwaps);

        List<String> listToSort = new ArrayList<>(list);
        CountingComparator listComp = new CountingComparator();
        long listSortTimeStart = System.nanoTime();
        listToSort.sort(listComp);
        long timeMicros = (System.nanoTime() - listSortTimeStart) / 1000;
        System.out.println("List: " + timeMicros + "µs; " + listComp.comparisons + " comparisons");

        CountingComparator treeComp = new CountingComparator();
        long treeSortTimeStart = System.nanoTime();
        TreeSet<String> tree = new TreeSet<>(treeComp);
        tree.addAll(list);
        long timeMicrosTree = (System.nanoTime() - treeSortTimeStart) / 1000;
        System.out.println("Tree: " + timeMicrosTree + "µs; " + treeComp.comparisons + " comparisons");

    }

    private static void swapRandom(List<String> list) {
        int first = ThreadLocalRandom.current().nextInt(list.size());
        int second = ThreadLocalRandom.current().nextInt(list.size());
        String atFirst = list.set(first, list.get(second));
        list.set(second, atFirst);
    }
```
gave these results:
```
Testing with 10; num swaps: 3
List: 222µs; 20 comparisons
Tree: 1096µs; 27 comparisons

Testing with 100; num swaps: 5
List: 260µs; 403 comparisons
Tree: 563µs; 785 comparisons

Testing with 1000; num swaps: 7
List: 6362µs; 1407 comparisons
Tree: 3387µs; 14411 comparisons

Testing with 10000; num swaps: 10
List: 658µs; 10886 comparisons
Tree: 7446µs; 211178 comparisons

Testing with 100000; num swaps: 12
List: 10479µs; 101214 comparisons
Tree: 42841µs; 2772129 comparisons

Testing with 1000000; num swaps: 14
List: 61146µs; 1002123 comparisons
Tree: 422747µs; 34378259 comparisons

Testing with 10000000; num swaps: 17
List: 866504µs; 10002954 comparisons
Tree: 9250023µs; 410639440 comparisons
```
So the difference becomes pretty significant >10k. It looks like the tree does far more comparisons that the list, so I imagine it would be fairly sensitive to the complexity of the comparison.

Reply:
Interesting. I'd guess that we typically have 1000 or fewer results. It's a 1ms improvement at that level, which in the context of the whole transaction is not very much at all. If we get beyond 10k results then we start to have other problems...

Reply:
If there isn't a downside, we should make the change. 1 ms not being an improvement can and should change as we work on transaction overheads to improve performance.